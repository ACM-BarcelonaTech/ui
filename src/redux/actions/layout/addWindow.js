import { updateLayout } from 'redux/actions/layout/index';
import _ from 'lodash';

const getSinglewindowConfig = (window) => ({
  first: window,
  second: '',
  splitPercentage: 100,
  direction: 'row',
});

const getMultipleWindowsConfig = (first, second) => ({
  first,
  second,
  splitPercentage: 60,
  direction: 'row',
});

const addWindow = (panel, window) => (dispatch, getState) => {
  const { layout } = getState();
  let newLayout;

  if (!layout.windows) {
    newLayout = {
      windows: getSinglewindowConfig(window || panel),
    };
  } else {
    newLayout = _.cloneDeep(layout);
    const allWindows = JSON.stringify(layout);

    // if the panel exists do not do anything
    if (allWindows.includes(panel)) return;

    // if the window exists - update it
    if (allWindows.includes(window)) {
      return dispatch(updateLayout(newLayout.windows, panel));
    }

    newLayout.windows.first = getMultipleWindowsConfig(layout.windows.first, window || panel);
  }
  return dispatch(updateLayout(newLayout.windows, panel));
};

export default addWindow;
