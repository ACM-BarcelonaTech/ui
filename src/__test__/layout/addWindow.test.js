import addWindow from 'redux/actions/layout/addWindow';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureStore([thunk]);

describe('add window', () => {
  it('adds a window to existing ones', () => {
    const state = {
      layout: {
        windows: {
          first: {
            first: 'samplewindow',
            second: 'secondFirstwindow',
            splitPercentage: 100,
            direction: 'row',
          },
          second: 'secondWindow',
          splitPercentage: 60,
          direction: 'row',
        },
      },
    };
    const store = mockStore(state);
    const newState = store.dispatch(addWindow('newPanel', 'newWindow'));
    expect(newState).toMatchSnapshot();
  });

  it('adds a new window if layout is empty', () => {
    const state = {
      layout: {},
    };
    const store = mockStore(state);
    const newState = store.dispatch(addWindow('newPanel', 'newWindow'));
    expect(newState).toMatchSnapshot();
  });
});
