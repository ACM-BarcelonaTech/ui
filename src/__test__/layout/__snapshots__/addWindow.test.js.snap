// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add window adds a new window if layout is empty 1`] = `
Object {
  "payload": Object {
    "panel": "newPanel",
    "windows": Object {
      "direction": "row",
      "first": "newWindow",
      "second": "",
      "splitPercentage": 100,
    },
  },
  "type": "UPDATE_LAYOUT",
}
`;

exports[`add window adds a window to existing ones 1`] = `
Object {
  "payload": Object {
    "panel": "newPanel",
    "windows": Object {
      "direction": "row",
      "first": Object {
        "direction": "row",
        "first": Object {
          "direction": "row",
          "first": "samplewindow",
          "second": "secondFirstwindow",
          "splitPercentage": 100,
        },
        "second": "newWindow",
        "splitPercentage": 60,
      },
      "second": "secondWindow",
      "splitPercentage": 60,
    },
  },
  "type": "UPDATE_LAYOUT",
}
`;
